### DISCUSSION TITLE
Simple way to configure route constraints.

### DISCUSSION THREAD

**Discussion by grounzero** [dotnet/yarp#729] [Q&A] [answered]
I am using `"Path": "{**catch-all}"` to allow all for a given endpoint.

I want to be able to allow all routes except `/log', '/health', '/metrics'`. 

All the examples I've come across seem to show how to configure this with controllers using code or attributes when I want to do it entirely in the `appsettings.json`

I tried using a regex.  `"Path": "{^(?!.*(log|health|metrics))}"` which appeared to work in a regex tester but putting it in the config gets me this. 

```
The route parameter name '^(?!.*(log|health|metrics))' is invalid. Route parameter names must be non-empty and cannot contain these characters: '{', '}', '/'. The '?' character marks a parameter as optional, and can occur only at the end of the parameter. The '*' character marks a parameter as catch-all, and can occur only at the start of the parameter.
```

 I am by no means a regex expert and am not even sure this is the way to go about this so thinking there must be a cleaner way to achieve this.

**Comment by grounzero** [comment]
Looks like I found what I want in https://docs.microsoft.com/en-us/aspnet/core/fundamentals/routing?view=aspnetcore-5.0#regular-expressions-in-constraints

**Comment by grounzero** [comment, marked as answer]
Rather than have ugly regex paths inside configuration files.  I added a middleware to ignore specific routes.  I quite like having the `"Path": "{**catch-all}"` with a catch all to ignore certain paths.  If there is a nicer way I would be very interested to know but for the moment this appears to do the job it needs to.

```c#
    public class IgnoredRouteMiddleware
    {
        readonly RequestDelegate _next;
        readonly IgnoredRouteOptions _options;

        public IgnoredRouteMiddleware(RequestDelegate next, IgnoredRouteOptions options) =>
            (_next, _options) = (next, options);

        public async Task Invoke(HttpContext context)
        {
            if (_options.Routes.Any(route =>
                context.Request.Path.HasValue &&
                context.Request.Path.StartsWithSegments(route, StringComparison.OrdinalIgnoreCase)))
            {
                context.Response.StatusCode = (int)_options.HttpStatusCode;
                return;
            }
            await _next.Invoke(context).ConfigureAwait(false);
        }
    }

    public class IgnoredRouteOptions
    {
        public IEnumerable<string> Routes { get; set; }
        public HttpStatusCode HttpStatusCode { get; set; }
    }
```

in `appsettings.json`

``` json
"IgnoredRoutes": {
    "Routes": [
      "/health",
      "/log",
      "/metrics",
      "/status"
    ],
    "HttpStatusCode": "Forbidden"
  }
```

**Reply by Tratcher** [reply]
Is there a reason you're using `Path.Value.Contains(route` rather than `Path.StartsWithSegments(route`? Would you really want to match those values anywhere in the path?

**Reply by grounzero** [reply]
Thanks.  I didn't know about `StartsWithSegments()`  I'll update now.

**Comment by Kevin-Andrew** [comment]
I believe this will suffer from a bypass vulnerability for a double slash path. For example, if I make a request to, `https://your.server.com//status`, (note the `//` before `status`) that will unfortunately not be matched by `context.Request.Path.StartsWithSegments(route, StringComparison.OrdinalIgnoreCase)`. And will therefore not be blocked/ignored, as you intended.

@Tratcher , are you still working on this project? Is this a known issue? Do you know of, or can you think of any reasonable workarounds to handle this vulnerability and accomplish the intended URL/path blocking that the author clearly intends.
